File: Service2/src/main/kotlin/com/exercise01/service2/services/StorageService.kt
package com.exercise01.service2.services

import com.exercise01.service2.models.LogEntry
import com.exercise01.service2.services.contracts.IStorageService
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.springframework.stereotype.Service
import java.io.File

private const val BASE_URL: String = "http://storage:1234/"
private const val V_STORAGE_LOG_PATH = "/app/data/log.txt"

@Service
class StorageService : IStorageService {
    private val client: OkHttpClient = OkHttpClient()
    private val objectMapper = jacksonObjectMapper()

    override fun addLog(message: String) {
        val logEntry = LogEntry(
            message = message,
            service = "Service2"
        )

        val jsonBody = objectMapper.writeValueAsString(logEntry)

        val request: Request = Request.Builder()
            .url(BASE_URL + "logs")
            .post(jsonBody.toRequestBody("application/json".toMediaType()))
            .build()

        try {
            client.newCall(request).execute()
            appendLogToVStorage(message)
        } catch (e: Exception) {
            println("Error adding log: ${e.message}")
        }
    }

    private fun appendLogToVStorage(message: String) {
        try {
            val logFile = File(V_STORAGE_LOG_PATH)

            logFile.parentFile?.mkdirs()

            if (!logFile.exists()) {
                logFile.createNewFile()
            }

            logFile.appendText(message + System.lineSeparator())
        } catch (e: Exception) {
            println("Failed to write log to volume storage: ${e.message}")
        }
    }
}

File: Service1/Service1/Services/StorageService.cs
using Service1.Models;
using Service1.Responses;
using Service1.Services.Contracts;

namespace Service1.Services;

public class StorageService : IStorageService
{
    private const string VStorageLogPath = "/app/data/log.txt";
    
    private readonly HttpClient _client;
    private readonly ILogger<StorageService> _logger;
    
    public StorageService(ILogger<StorageService> logger)
    {
        _client = new HttpClient();
        _client.BaseAddress = new Uri("http://storage:1234/");
        _client.DefaultRequestHeaders.Accept.Clear();

        _logger = logger;
    }
    
    public async Task<string?> GetLog()
    {
        try
        {
            var response = await _client.GetAsync("logs");
            response.EnsureSuccessStatusCode();
            
            var logEntries = await response.Content.ReadFromJsonAsync<GetLogsResponse>();

            return logEntries is null ? null : string.Join(Environment.NewLine, logEntries.Logs.Select(l => l.Message));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting log");
            return null;
        }
    }

    public async Task AddLog(string message)
    {
        try
        {
            var logEntry = new LogEntry
            {
                Message = message,
                Service = "Service1"
            };

            var response = await _client.PostAsJsonAsync("logs", logEntry);
            response.EnsureSuccessStatusCode();
            
            AppendLogToVStorage(message);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error adding log");
        }
    }

    private void AppendLogToVStorage(string message)
    {
        try
        {
            var directory = Path.GetDirectoryName(VStorageLogPath);
            if (!Directory.Exists(directory))
                Directory.CreateDirectory(directory!);
            
            if (!File.Exists(VStorageLogPath))
                File.Create(VStorageLogPath).Close();
            
            File.AppendAllText(VStorageLogPath, message + Environment.NewLine);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to write log to volume storage");
        }
    }
}